File: common/src/main/java/juuxel/adorn/block/OxidizableCopperPipeBlock.java
package juuxel.adorn.block;

import juuxel.adorn.platform.BlockBridge;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.util.RandomSource;
import net.minecraft.world.level.block.WeatheringCopper;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.BlockState;

public final class OxidizableCopperPipeBlock extends CopperPipeBlock implements WeatheringCopper {
    private final WeatherState oxidationLevel;

    public OxidizableCopperPipeBlock(WeatheringCopper.WeatherState oxidationLevel, BlockBehaviour.Properties settings) {
        super(settings);
        this.oxidationLevel = oxidationLevel;
    }

    @Override
    public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
        changeOverTime(state, world, pos, random);
    }

    @Override
    public boolean isRandomlyTicking(BlockState state) {
        return BlockBridge.get().hasNextOxidationLevelForTicking(state);
    }

    @Override
    public WeatherState getAge() {
        return oxidationLevel;
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/OxidizableCopperPipeBlock.kt
@file:Suppress("OVERRIDE_DEPRECATION")

package juuxel.adorn.block

import net.minecraft.block.BlockState
import net.minecraft.block.Oxidizable
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.random.Random

class OxidizableCopperPipeBlock(
    private val oxidationLevel: Oxidizable.OxidationLevel,
    settings: Settings
) : CopperPipeBlock(settings), Oxidizable {
    override fun randomTick(state: BlockState, world: ServerWorld, pos: BlockPos, random: Random) {
        tickDegradation(state, world, pos, random)
    }

    override fun hasRandomTicks(state: BlockState): Boolean =
        Oxidizable.getIncreasedOxidationBlock(state.block).isPresent

    override fun getDegradationLevel(): Oxidizable.OxidationLevel =
        oxidationLevel
}
